evm/src/cpu/kernel/asm/secp256k1/curve_add.asm

```diff
@@ -0,0 +1,338 @@
+    MULMOD
+    // stack: lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    DUP7
+    // stack: y1, lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    SWAP1
+    // stack: lambda * (x1 - x2), y1, x2, lambda, x0, y0, x1, y1, retdest
+    %submod_secp_base
+    // stack: y2, x2, lambda, x0, y0, x1, y1, retdest
+
+    // Return x2,y2
+    SWAP5
+    // stack: x1, x2, lambda, x0, y0, y2, y1, retdest
+    POP
+    // stack: x2, lambda, x0, y0, y2, y1, retdest
+    SWAP5
+    // stack: y1, lambda, x0, y0, y2, x2, retdest
+    %pop4
+    // stack: y2, x2, retdest
+    SWAP2
+    // stack: retdest, x2, y2
+    JUMP
+
+// Secp256k1 elliptic curve addition.
+// Assumption: (x0,y0) and (x1,y1) are valid points and x0 == x1
+ec_add_equal_first_coord:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest with x0 == x1
+
+    // Check if the points are equal
+    DUP2
+    // stack: y0, x0, y0, x1, y1, retdest
+    DUP5
+    // stack: y1, y0, x0, y0, x1, y1, retdest
+    EQ
+    // stack: y1 == y0, x0, y0, x1, y1, retdest
+    %jumpi(ec_add_equal_points)
+    // stack: x0, y0, x1, y1, retdest
+
+    // Otherwise, one is the negation of the other so we can return (0,0).
+    %pop4
+    // stack: retdest
+    PUSH 0
+    // stack: 0, retdest
+    PUSH 0
+    // stack: 0, 0, retdest
+    SWAP2
+    // stack: retdest, 0, 0
+    JUMP
+
+
+// Secp256k1 elliptic curve addition.
+// Assumption: x0 == x1 and y0 == y1
+// Standard doubling formula.
+ec_add_equal_points:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+
+    // Compute lambda = 3/2 * x0^2 / y0
+    %secp_base
+    // stack: N, x0, y0, x1, y1, retdest
+    %secp_base
+    // stack: N, N, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: x0, N, N, x0, y0, x1, y1, retdest
+    DUP1
+    // stack: x0, x0, N, N, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: x0^2, N, x0, y0, x1, y1, retdest with
+    PUSH 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffff7ffffe19 // 3/2 in the base field
+    // stack: 3/2, x0^2, N, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: 3/2 * x0^2, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: y0, 3/2 * x0^2, x0, y0, x1, y1, retdest
+    %moddiv_secp_base
+    // stack: lambda, x0, y0, x1, y1, retdest
+    %jump(ec_add_valid_points_with_lambda)
+
+// BN254 elliptic curve doubling.
+// Assumption: (x0,y0) is a valid point.
+// Standard doubling formula.
+global ec_double_secp:
+    JUMPDEST
+    // stack: x0, y0, retdest
+    DUP2
+    // stack: y0, x0, y0, retdest
+    DUP2
+    // stack: x0, y0, x0, y0, retdest
+    %jump(ec_add_equal_points)
+
+// Push the order of the Secp256k1 scalar field.
+%macro secp_base
+    PUSH 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f
+%endmacro
+
+%macro submod_secp_base
+    // stack: x, y
+    SWAP1
+    // stack: y, x
+    DUP2
+    // stack: x, y, x
+    SUB
+    // stack: x - y, x
+    DUP1
+    // stack: x - y, x - y, x
+    SWAP2
+    // stack: x, x - y, x - y
+    LT
```

Comment:
This makes sense (after seeing the method in @nbgl's SIMD code) but please add a brief note, since it's not obvious that this is an underflow check